meson.build

```diff
@@ -0,0 +1,68 @@
+# SPDX-License-Identifier: LGPL-2.1 OR BSD-2-Clause
+project('libbpf', 'c',
+        version : '0.0.2',
+        license : 'LGPL-2.1 OR BSD-2-Clause')
+
+libbpf_src_dir = './'
+
+libbpf_sources = files(run_command('find',
+        ['@0@/src'.format(libbpf_src_dir),
+        '-type',
+        'f',
+        '-name',
+        '*.h',
+        '-o',
+        '-name',
+        '*.c']).stdout().split())
+
+libbpf_headers = files(
+        join_paths(libbpf_src_dir, 'src/bpf.h'),
+        join_paths(libbpf_src_dir, 'src/btf.h'),
+        join_paths(libbpf_src_dir, 'src/libbpf.h'))
+
+
+feature_rellocarray = run_command(join_paths(libbpf_src_dir, 'scripts/check-reallocarray.sh'))
+
+libbpf_c_args = [
+              '-g',
+              '-O2',
+              '-Werror',
+              '-Wall',
+              '-Wno-pointer-arith',
+              '-Wno-sign-compare'
+]
+
+if feature_rellocarray.stdout().strip() != ''
+        libbpf_c_args += '-DCOMPAT_NEED_REALLOCARRAY'
+endif
+
+libbpf_includes = include_directories(
+        join_paths(libbpf_src_dir, 'include'),
+        join_paths(libbpf_src_dir, 'include/uapi'))
+
+
+# bpf_includes are required to include bpf.h, btf.h, libbpf.h
+bpf_includes = include_directories(
+        join_paths(libbpf_src_dir, 'src'))
+
+libbpf_static = static_library(
+        'bpf',
+        libbpf_sources,
+        include_directories : libbpf_includes,
+        install : false,
+        c_args : libbpf_c_args)
+
+libbpf_static_dep = declare_dependency(link_with : libbpf_static)
+
+libbpf_map = join_paths(libbpf_src_dir, 'src/libbpf.map')
+
+libbpf_c_args += ['-fPIC', '-fvisibility=hidden']
+
+libbpf_shared = shared_library(
+        'bpf',
+        include_directories : libbpf_includes,
+        link_depends : libbpf_map,
+        install : false,
+        c_args : libbpf_c_args)
```

Comment:
I see from documentation that `libbpf_shared` supports `version` and `soversion` and uses reasonable values by default, but the result it provides with this configuration differs from what `Makefile` does.

I built libbpf both with meson in `build/` directory and with `Makefile` in `src/`.

`SONAME` should be fixed:
```
% diff -u <(readelf -d src/libbpf.so  | grep SONAME) <(readelf -d build/libbpf.so  | grep SONAME)
--- /dev/fd/63  2019-04-10 12:57:06.593038040 -0700
+++ /dev/fd/62  2019-04-10 12:57:06.594038036 -0700
@@ -1 +1 @@
- 0x000000000000000e (SONAME)             Library soname: [libbpf.so.0]
+ 0x000000000000000e (SONAME)             Library soname: [libbpf.so]
```

libelf should be linked:
```
% diff -u <(ldd src/libbpf.so | cut -d" " -f 3 | grep lib) <(ldd build/libbpf.so | cut -d" " -f 3 | grep lib)
--- /dev/fd/63  2019-04-10 13:03:56.108390558 -0700
+++ /dev/fd/62  2019-04-10 13:03:56.109390554 -0700
@@ -1,3 +1 @@
-/lib64/libelf.so.1
 /lib64/libc.so.6
-/lib64/libz.so.1
```

Versioning script `libbpf.map` should be used (not it seems it's not):
```
diff -u <(readelf -s --wide src/libbpf.so | grep -Eo '[^ ]+@LIBBPF_.*' | cut -d@ -f 3 | sort | uniq -c) <(readelf -s --wide build/libbpf.so | grep -Eo '[^ ]+@LIBBPF_.*' | cut -d@ -f 3 | sort | uniq -c)
--- /dev/fd/63  2019-04-10 13:05:29.687021896 -0700
+++ /dev/fd/62  2019-04-10 13:05:29.688021892 -0700
@@ -1,2 +0,0 @@
-    121 LIBBPF_0.0.1
-     29 LIBBPF_0.0.2
```

Also I"m not sure how specifically a project will use shared lib, but it might make sense to provide version symlinks same way `Makefile` does it now:
```
% ls -al src/libbpf.so*
lrwxrwxrwx. 1 rdna users     11 Apr 10 12:55 src/libbpf.so -> libbpf.so.0
lrwxrwxrwx. 1 rdna users     15 Apr 10 12:55 src/libbpf.so.0 -> libbpf.so.0.0.2
-rwxr-xr-x. 1 rdna users 377632 Apr 10 12:55 src/libbpf.so.0.0.2
```

See `Makefile` for details on how it does all these things.

Reply:
To find the libelf dependency in a way that's compatible with both a system with pkg-config and without, this snippet can be used:

```
libelf = dependency('libelf', required: false)
if not libelf.found()
        libelf = cc.find_library('elf', required: true)
endif

...

shared_library(...
    dependencies: libelf,
...
```

Reply:
@bluca It looks Meson handles this internally on `master` https://github.com/mesonbuild/meson/blob/master/mesonbuild/dependencies/base.py#L572-L592

I'm a bit surprised that Meson doesn't handle this internally, studying the code just in case if it does.

Reply:
@rdna 
Discrepancies you mentioned are gone: https://github.com/libbpf/libbpf/commit/86ed4a6cfdc41886d8cff2f7dd120566759f92bb

Also` install` part is similar to `make install` with the difference that there is no check for `BUILD_STATIC_ONLY` envar, so both `*.a` and `*.so` are installed. It doesn't matter for systemd use case.

Reply:
It doesn't fall back automatically unfortunately and it's mightily annoying, there's an issue open on their repo that asks for the feature but it's not done yet